# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [13]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor,AdaBoostRegressor, GradientBoostingRegressor

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

In [3]:
spaceship.isnull().sum()

PassengerId       0
HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [4]:
spaceship = spaceship.dropna().reset_index()

In [7]:
for i in range(0, len(spaceship)):
    spaceship.iloc[i, 4] = spaceship['Cabin'][i][0]

In [8]:
columns_to_drop = ['PassengerId', 'Name']
spaceship.drop(columns=columns_to_drop, inplace=True)

In [9]:
spaceship_dummies = pd.get_dummies(spaceship, columns = ['HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'VIP', 'Transported'])

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [10]:
features = spaceship_dummies
target = spaceship_dummies['Spa']  # Reemplaza 'TargetVariable' con el nombre de la columna que contiene el target


**Perform Train Test Split**

In [11]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.20)

In [15]:
#primero normalizamos

In [16]:
normalizer = MinMaxScaler()

normalizer.fit(X_train)

MinMaxScaler()

In [17]:
X_train_norm = normalizer.transform(X_train)

X_test_norm = normalizer.transform(X_test)

In [18]:
X_train_norm = pd.DataFrame(X_train_norm, columns = X_train.columns)
X_train_norm.head()

index       Age  RoomService  FoodCourt  ShoppingMall       Spa  \
0  0.501553  0.569620     0.006552   0.002113      0.000245  0.031819   
1  0.386952  0.126582     0.000000   0.000000      0.000000  0.000000   
2  0.078587  0.329114     0.158972   0.000067      0.000000  0.001339   
3  0.508572  0.341772     0.000000   0.021132      0.000000  0.003169   
4  0.752272  0.582278     0.000000   0.000000      0.000000  0.000000   

     VRDeck  HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  \
0  0.001328               1.0                0.0              0.0  ...   
1  0.000000               1.0                0.0              0.0  ...   
2  0.000000               1.0                0.0              0.0  ...   
3  0.000000               1.0                0.0              0.0  ...   
4  0.000000               0.0                1.0              0.0  ...   

   Cabin_F  Cabin_G  Cabin_T  Destination_55 Cancri e  \
0      1.0      0.0      0.0                      0.0   
1      0.0      1.0      0.0                      1.0   
2      0.0      0.0      0.0                      0.0   
3      1.0      0.0      0.0                      1.0   
4      0.0      0.0      0.0                      0.0   

   Destination_PSO J318.5-22  Destination_TRAPPIST-1e  VIP_False  VIP_True  \
0                        0.0                      1.0        1.0       0.0   
1                        0.0                      0.0        1.0       0.0   
2                        0.0                      1.0        1.0       0.0   
3                        0.0                      0.0        1.0       0.0   
4                        0.0                      1.0        1.0       0.0   

   Transported_False  Transported_True  
0                1.0               0.0  
1                1.0               0.0  
2                1.0               0.0  
3                0.0               1.0  
4                0.0               1.0  

[5 rows x 27 columns]

In [19]:
X_test_norm = pd.DataFrame(X_test_norm, columns = X_test.columns)
X_test_norm.head()

index       Age  RoomService  FoodCourt  ShoppingMall       Spa  VRDeck  \
0  0.623288  0.430380      0.05756   0.000000      0.000000  0.000000     0.0   
1  0.331953  0.316456      0.00000   0.000000      0.000000  0.000000     0.0   
2  0.767230  0.253165      0.00000   0.000000      0.190158  0.005355     0.0   
3  0.622138  0.772152      0.00000   0.000000      0.000000  0.000000     0.0   
4  0.870096  0.227848      0.05998   0.004495      0.000571  0.000000     0.0   

   HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  ...  Cabin_F  \
0               1.0                0.0              0.0  ...      1.0   
1               0.0                0.0              1.0  ...      0.0   
2               1.0                0.0              0.0  ...      1.0   
3               0.0                1.0              0.0  ...      0.0   
4               1.0                0.0              0.0  ...      1.0   

   Cabin_G  Cabin_T  Destination_55 Cancri e  Destination_PSO J318.5-22  \
0      0.0      0.0                      0.0                        1.0   
1      0.0      0.0                      0.0                        0.0   
2      0.0      0.0                      0.0                        0.0   
3      0.0      0.0                      1.0                        0.0   
4      0.0      0.0                      0.0                        0.0   

   Destination_TRAPPIST-1e  VIP_False  VIP_True  Transported_False  \
0                      0.0        1.0       0.0                1.0   
1                      1.0        1.0       0.0                0.0   
2                      1.0        1.0       0.0                0.0   
3                      0.0        1.0       0.0                0.0   
4                      1.0        1.0       0.0                1.0   

   Transported_True  
0               0.0  
1               1.0  
2               1.0  
3               1.0  
4               0.0  

[5 rows x 27 columns]

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [81]:
bagging_reg = BaggingRegressor(DecisionTreeRegressor(max_depth=20),
                               n_estimators=150,
                               max_samples = 1400)

In [85]:
#entreno el bagging model con la data normalizada

bagging_reg.fit(X_train_norm, y_train)

BaggingRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                 max_samples=1400, n_estimators=150)

In [86]:
#evaluo el performance del modelo

pred = bagging_reg.predict(X_test_norm)

bagging_MAE = mean_absolute_error(pred, y_test)
bagging_RMSE = mean_squared_error(pred, y_test, squared=False)
bagging_R2 = bagging_reg.score(X_test_norm, y_test)

bagging_MAE, bagging_RMSE, bagging_R2

(2.081215330307614, 18.159097568665715, 0.9996393087199626)

- Random Forests

In [96]:
#hago un random forest 

forest = RandomForestRegressor(n_estimators=100,
                             max_depth=20)

In [97]:
#entreno el modelo

forest.fit(X_train_norm, y_train)

RandomForestRegressor(max_depth=20)

In [98]:
#evaluo el modelo

pred = forest.predict(X_test_norm)

random_forest_MAE = mean_absolute_error(pred, y_test)
random_forest_RMSE = mean_squared_error(pred, y_test, squared=False)
random_forest_R2 = forest.score(X_test_norm, y_test)

random_forest_MAE, random_forest_RMSE, random_forest_R2

(1.5435098335854762, 14.182558208495644, 0.9997799829850033)

- Gradient Boosting

In [25]:
#inicio un gradient boost modelo

gb_reg = GradientBoostingRegressor(max_depth=20,
                                   n_estimators=100)

In [26]:
#entreno el modelo

gb_reg.fit(X_train_norm, y_train)

GradientBoostingRegressor(max_depth=20)

In [33]:
#evaluo el modelo

pred = gb_reg.predict(X_test_norm)

gradient_boosting_MAE = mean_absolute_error(pred, y_test)
gradient_boosting_RMSE = mean_squared_error(pred, y_test, squared=False)
gradient_boosting_R2 = gb_reg.score(X_test_norm, y_test)

gradient_boosting_MAE, gradient_boosting_RMSE, gradient_boosting_R2

(2.44667697280698, 26.155727442498907, 0.9992516914194737)

- Adaptive Boosting

In [28]:
#inicio un modelo adaboost

ada_reg = AdaBoostRegressor(DecisionTreeRegressor(max_depth=20),
                            n_estimators=100)

In [29]:
#entreno el modelo

ada_reg.fit(X_train_norm, y_train)

AdaBoostRegressor(estimator=DecisionTreeRegressor(max_depth=20),
                  n_estimators=100)

In [32]:
#evaluo el modelo

pred = ada_reg.predict(X_test_norm)

adaptive_boosting_MAE = mean_absolute_error(pred, y_test)
adaptive_boosting_RMSE = mean_squared_error(pred, y_test, squared=False)
adaptive_boosting_R2 = ada_reg.score(X_test_norm, y_test)

adaptive_boosting_MAE, adaptive_boosting_RMSE, adaptive_boosting_R2

(1.8229954614220878, 22.897423039409663, 0.9994265172820666)

Which model is the best and why?

In [103]:
print('Adaboost model results. MAE:', adaptive_boosting_MAE, 'RMSE:', adaptive_boosting_RMSE, 'R2:', adaptive_boosting_R2)
print('GradientBoost model results. MAE:', gradient_boosting_MAE, 'RMSE:', gradient_boosting_RMSE, 'R2:', gradient_boosting_R2)
print('Random Forest model. MAE:', random_forest_MAE, 'RMSE:', random_forest_RMSE, 'R2:', random_forest_R2)
print('Bagging model results. MAE:', bagging_MAE, 'RMSE:', bagging_RMSE, 'R2:', bagging_R2)

Adaboost model results. MAE: 1.8229954614220878 RMSE: 22.897423039409663 R2: 0.9994265172820666
GradientBoost model results. MAE: 2.44667697280698 RMSE: 26.155727442498907 R2: 0.9992516914194737
Random Forest model. MAE: 1.5435098335854762 RMSE: 14.182558208495644 R2: 0.9997799829850033
Bagging model results. MAE: 2.081215330307614 RMSE: 18.159097568665715 R2: 0.9996393087199626


# Conclusiones:

El modelo que mejor funciona es el Random Forest ya que reduce el número de errores y aumenta el R2 simultáneamente en comparación con los otros modelos analizados:

- Adaboost model results: 1.8229954614220878 22.897423039409663 0.9994265172820666

- GradientBoost model results: 2.44667697280698 26.155727442498907 0.9992516914194737

- Random Forest model: 1.5435098335854762 14.182558208495644 0.9997799829850033

- Bagging model results: 2.081215330307614 18.159097568665715 0.9996393087199626